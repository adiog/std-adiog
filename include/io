// This file is a part of std-adiog project.
// Copyright (c) 2016 Aleksander Gajewski <adiog@brainfuck.pl>.

#ifndef _IO
#define _IO

#include <_adg_prettyprint.h>

#include <_adg_iostream.h>

#endif
