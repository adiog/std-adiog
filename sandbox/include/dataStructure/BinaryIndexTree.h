// This file is a part of cpp-training project.
// Copyright (c) 2016 Aleksander Gajewski <adiog@brainfuck.pl>.

#ifndef CPP_TRAINING_BINARYINDEXTREE_H
#define CPP_TRAINING_BINARYINDEXTREE_H

#include <algorithm>
#include <vector>

namespace dataStructure {
class BinaryIndexTree {
  public:
    explicit BinaryIndexTree(int size) : v(size + 1, 0) {}
    explicit BinaryIndexTree(const std::vector<int> &in) : v(in.size() + 1, 0) {
        for (int i = 0; i < in.size(); i++) {
            add(i, in[i]);
        }
    }

    int get(int i) const {
        int sum = 0;
        while (i > 0) {
            sum += v[i];
            i -= i & -i;
        }
        return sum;
    }

    void add(int i, int k) {
        i++;
        while (i < v.size()) {
            v[i] += k;
            i += i & -i;
        }
    }

    std::vector<int> getValues() const {
        std::vector<int> ret(v.size());
        int i = 0;
        std::generate(ret.begin(), ret.end(), [&]() { return this->get(i++); });
        return ret;
    }

    const std::vector<int> &getRepresentation() const { return v; }

  private:
    std::vector<int> v;
};
}

/*
 * std::ostream &operator<<(
    std::ostream &os, const dataStructure::BinaryIndexTree &binaryIndexTree) {
    os << binaryIndexTree.getRepresentation();

    return os;
}
*/
#endif  // CPP_TRAINING_BINARYINDEXTREE_H
