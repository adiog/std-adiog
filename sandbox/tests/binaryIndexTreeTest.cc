#include <io>
#include "dataStructure/BinaryIndexTree.h"

using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> v(n);
    cin >> v;
    dataStructure::BinaryIndexTree bit(v);
    cout << bit << endl;
    cout << bit.getValues() << endl;
    return 0;
}