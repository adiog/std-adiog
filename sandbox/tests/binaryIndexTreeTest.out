[0, 2, 5, 4, 5, 9]
[0, 2, 5, 9, 5, 14]
